### PR TITLE
introduce lzw encode/decode

### PR DESCRIPTION
**Pull request by ykuzmenko73**
compressed = lzw_encode(sourceStr);
decompressed = lzw_decode(compressed);

for wigi strings 70k size:
	compressed = lzw_encode(sourceStr);
	decompressed = lzw_decode(compressed);
	println(i2s(strlen(sourceStr)) + " -> " + i2s(length(compressed)) + " -> " + i2s(strlen(decompressed)));

it gives
70601 -> 10404 -> 70601

### CODE REVIEW

File: lib/text/lzw.flow [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+lzw_encode(s : string) -> [int] {
+	dict : ref Tree<[int], int> = ref makeTree();
+	out : ref List<int> = ref makeList();
+	phrase : ref [int]= ref [getCharCodeAt(s, 0)];
+	code = ref 65536;
+	fori(1, strlen(s)-1, \i -> {
+		currChar = getCharCodeAt(s, i);
+		newPrase = arrayPush(^phrase, currChar);
+		if (isSome(lookupTree(^dict, newPrase))) {
+			refArrayPush(phrase,  currChar);
```
Comment (yegordovganich) [inline lib/text/lzw.flow]:
```suggestion
			phrase := newPhrase;
```

File: lib/text/lzw.flow [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+			);
+			dict := setTree(^dict, newPrase, ^code);
+			code := ^code + 1;
+			phrase := [currChar];
+		}
+	});
+
+	out := Cons(
+		if (length(^phrase) > 1) {
+			lookupTreeDef(^dict, ^phrase, 0);
```
Comment (yegordovganich) [inline lib/text/lzw.flow:41]: make it as lambda to call only pushToList(out, phrase); - instead of copy past

File: lib/text/lzw.flow [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	// Decompress an LZW-encoded string
+	lzw_decode(data : [int]) -> string;
+}
+
+lzw_encode(s : string) -> [int] {
+	dict : ref Tree<[int], int> = ref makeTree();
+	out : ref List<int> = ref makeList();
+	phrase : ref [int]= ref [getCharCodeAt(s, 0)];
+	code = ref 65536;
+	fori(1, strlen(s)-1, \i -> {
```
Comment (yegordovganich) [inline lib/text/lzw.flow:19]: use foldi to keep it in functional style with some nice struct inside

File: lib/text/lzw.flow [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+	currChar : ref int = ref data[0];
+	out : ref List<[int]> = ref makeList1([^currChar]);
+	oldPhrase : ref [int] = ref [(^currChar)];
+	code = ref 65536;
+	fori (1, length(data)-1, \i -> {
+		currCode = data[i];
+		phrase = if (currCode < 65536) {
+			[data[i]];
+		} else {
+		   lookupTreeDef(^dict, currCode, arrayPush(^oldPhrase, ^currChar));
```
Comment (yegordovganich) [inline lib/text/lzw.flow]: add lookupTreeDefLazy based on eitherFn to call here:
`
lookupTreeDefLazy(^dict, currCode, \ -> arrayPush(^oldPhrase, ^currChar));
`

**Comment by ykuzmenko73** [comment]
@alstrup , I'm not sure about the licensing of LZW algo. Who could tell definitely if we are able to implement it?

**Comment by ykuzmenko73** [comment]
I will check if encoding could help.

File: lib/text/lzw.flow (hunk 3 above)
Comment (ykuzmenko73) [reply lib/text/lzw.flow:19]: I'm afraid it will increase the load on the GC :-(

File: lib/text/lzw.flow [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	currChar : ref int = ref data[0];
+	out : ref List<[int]> = ref makeList1([^currChar]);
+	oldPhrase : ref [int] = ref [(^currChar)];
+	code = ref 65536;
+	fori (1, length(data)-1, \i -> {
+		currCode = data[i];
+		phrase = if (currCode < 65536) {
+			[data[i]];
+		} else {
+			lookupTreeDef(^dict, currCode, arrayPush(^oldPhrase, ^currChar));
```
Comment (brawaga) [inline lib/text/lzw.flow:63]: It will output data[0] twice to phrase at first cycle pass?

File: lib/text/lzw.flow (hunk 3 above)
Comment (yegordovganich) [reply lib/text/lzw.flow:19]: OK, probably you are right

File: lib/text/lzw.flow (hunk 5 above)
Comment (ykuzmenko73) [reply lib/text/lzw.flow]: the cycle starts from 1, not 0

File: lib/text/lzw.flow [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+			phrase := [currChar];
+		}
+		lzw_encode_rec(s, sLength, i + 1, code, phrase, dict, storeResult)
+	} else {
+		storeResult()
+	}
+}
+
+lzw_encode(s : string) -> string {
+	dict : ref Tree<[int], int> = ref makeTree();
```
[the remark this answers is not in the document]
Comment (ykuzmenko73) [reply lib/text/lzw.flow]: indeed. please address this issue after merge

**Pull request merged**